engine/timer.cpp

```diff
@@ -0,0 +1,50 @@
+#include "timer.h"
+
+#include <spdlog/spdlog.h>
+
+#include <algorithm>
+#include <chrono>
+#include <iostream>
+#include <memory>
+
+namespace pong
+{
+
+static constexpr int MS_TO_US = 1000;
+static constexpr int S_TO_US = 1000000;
+
+float Timer::frameTime = 1.0f;
+
+void Timer::HandleTimerCallbacks()
+{
+    const auto timeWaited = std::chrono::duration_cast<std::chrono::microseconds>(std::chrono::system_clock::now() - mLastTime);
+    mLastTime = std::chrono::system_clock::now();
+
+    Timer::frameTime = static_cast<float>(timeWaited.count()) / static_cast<float>(S_TO_US);
+
+    const int timeWaitedUs = static_cast<int>(timeWaited.count());
+    for (auto& timerRequest : mActiveTimers)
+    {
+        timerRequest->mTimeoutUs -= timeWaitedUs;
+        if (timerRequest->mTimeoutUs <= 0)
+        {
+            timerRequest->mCallback();
+        }
+    }
+
+    mActiveTimers.erase(std::remove_if(mActiveTimers.begin(), mActiveTimers.end(), [](const auto& timerRequest) {
+        return timerRequest->mTimeoutUs <= 0;
+    }), mActiveTimers.end());
+}
+
+void Timer::AddTimer(int gameObjectId, int timeoutMs, std::function<void()> callback)
+{
+    // Interface uses ms but internally keeps track using us. This is just in case the
+    // game's framerate is uncapped. It is possible that {timeWaited} is less then 1ms
+    // and comes out to 0 then the timer will not proceed.
+    const int timeoutUs = timeoutMs * MS_TO_US;
+    auto timerRequest = std::make_shared<TimerRequest>(gameObjectId, timeoutUs, callback);
+    mActiveTimers.push_back(timerRequest);
```

Comment:
I don't know if you need to use a shared pointer here? You may just be able to push the TimerRequest object itself on there with no pointer wrapping it?

Reply:
If you do that, may be a good opportunity to use `std::move` here, or even `mActiveTimers.emplace_back`

Homework question: What does `emplace_back` do, and why would one consider it?

Reply:
Yeah that was a theory I had at one point, whether they need to be shared_ptr or not. I think I was just used to our old SplArray that forced everything to be allocated or it wouldn't work. Sometimes I just forget to test my theory 😁. I think I also had made this a shared_ptr because I was originally considering a separate thread for this timer class haha.

I looked at emplace_back briefly when I was setting up the objects in Pong::Init(), but I don't think I was able to because I needed to call OnStart() on the object after putting it in the array. Thats pretty cool to avoid the temporary object.
It seems like you would always try to use emplace_back though. Would there be any reason to not use emplace_back if you were able to?

Reply:
emplace_back is generally the way to do things if you are building the object then immediately adding it to the collection! I don't think there is any reason not to use it in that case....

push_back is when you have something already constructed and you want to append